text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,972 @@
+associated type).
+
+# Detailed design
+
+This section dives into detail on the proposed type rules.
+
+### A little type grammar
+
+We extend the type grammar from [RFC 192] with projections and slice
+types:
+
+    T = scalar (i32, u32, ...)        // Boring stuff
+      | X                             // Type variable
+      | Id<P0..Pn>                    // Nominal type (struct, enum)
+      | &r T                          // Reference (mut doesn't matter here)
+      | O0..On+r                      // Object type
+      | [T]                           // Slice type
+      | for<r..> fn(T1..Tn) -> T0     // Function pointer
+      | <P0 as Trait<P1..Pn>>::Id     // Projection
+    P = r                             // Region name
+      | T                             // Type
+    O = for<r..> TraitId<P1..Pn>      // Object type fragment
+    r = 'x                            // Region name
+    
+We'll use this to describe the rules in detail.
+    
+### Syntactic definition of the outlives relation
+
+The outlives relation is defined in purely syntactic terms as follows.
+These are inference rules written in a primitive ASCII notation. :) As
+part of defining the outlives relation, we need to track the set of
+lifetimes that are bound within the type we are looking at.  Let's
+call that set `R=<r0..rn>`. Initially, this set `R` is empty, but it
+will grow as we traverse through types like fns or objects, which can
+bind region names.
+
+#### Simple outlives rules
+
+Here are the rules covering the simple cases, where no type parameters
+or projections are involved:
+
+    OutlivesScalar:
+      --------------------------------------------------
+      R ⊢ scalar: 'a
+
+    OutlivesNominalType:
+      ∀i. Pi: 'a
+      --------------------------------------------------
+      R ⊢ Id<P0..Pn>: 'a
+
+    OutlivesReference:
+      R ⊢ 'x: 'a
+      R ⊢ T: 'a
+      --------------------------------------------------
+      R ⊢ &'x T: 'a
+
+    OutlivesObject:
+      ∀i. R ⊢ Oi: 'a
+      R ⊢ 'x: 'a
+      --------------------------------------------------
+      R ⊢ O0..On+'x: 'a
+
+    OutlivesFunction:
+      ∀i. R,r.. ⊢ Ti: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> fn(T1..Tn) -> T0
+
+    OutlivesTraitRef:
+      ∀i. R,r.. ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ for<r..> TraitId<P0..Pn>: 'a      
+      
+#### Outlives for lifetimes
+
+The outlives relation for lifetimes depends on whether the lifetime in
+question was bound within a type or not. In the usual case, we decide
+the relationship between two lifetimes by consulting the environment.
+Lifetimes representing scopes within the current fn have a
+relationship derived from the code itself, lifetime parameters have
+relationships defined by where-clauses and implied bounds:
+
+      'x not in R
+      ('x: 'a) in Env
+      --------------------------------------------------
+      R ⊢ 'x: 'a
```

Comment:
Env being "the union of all in scope where clauses and implied bounds"? Unclear.

Also unclear what the significance of `'x not in R` is? Are you just saying that R := R U Env, or that R only inherits bounds for otherwise unbound lifetimes?